contracts/AbstractConference.sol

```diff
@@ -70,15 +75,17 @@ contract AbstractConference is Conference, GroupAdmin {
      * @param _limitOfParticipants The number of participant. The default is set to 20. The number can be changed by the owner of the event.
      * @param _coolingPeriod The period participants should withdraw their deposit after the event ends. After the cooling period, the event owner can claim the remining deposits.
      * @param _owner The owner of the event
-     * @param _clearFee the fee for _clearAndSend function in per-mille (e.g. _clearFee = 10 means 1% fees, _clearFee = 1 means 0.1% fees) 
+     * @param _clearFee the fee for _clearAndSend function in per-mille (e.g. _clearFee = 10 means 1% fees, _clearFee = 1 means 0.1% fees)
+     * @param _baseTokenUri The base URI used for tokens.
      */
     constructor (
         string memory _name,
         uint256 _deposit,
         uint256 _limitOfParticipants,
         uint256 _coolingPeriod,
         address payable _owner,
-        uint256 _clearFee
+        uint256 _clearFee,
+        string memory _baseTokenUri
```

Comment:
Rather than passing _baseTokenUri at constructor, can you not save the deployer address and reference the baseTokenUri via Deployer.

For example, define deployer variable

```
Deployer  public deployer;
```

Then set at constructor

```
deployer = msg.sender
```

then `tokenURI` becomes like this.

```
   function tokenURI(uint256 tokenId) public view returns (string memory) { 
        return string(abi.encodePacked(deployer.baseTokenURI(), Utils.addr2str(address(this)), '/', _tokenURI(tokenId)));
    }
```

Most variables(deposit, cooling period, etc) has to be set at the constructor to make the variable unique to the event.
However, the only time we change baseTokenURI is when we change the metadata server to some other domain so it should be referred to the latest baseTokenURI, rather than passed by the constructor and fixed.

This also has the benefit of reducing changes on frontend/backend code as we are only adding ERC721 related functions, not changing existing interface.

Reply:
Setting `deployer = msg.sender` at the constructor will not work because the sender in this case is either `EthDeployer` or `ERC20Deployer` and not `Deployer`. Thus we can't call `deployer.baseTokenURI()` as suggested.

Maybe passing the address of `Deployer` instead of `_baseTokenUri `  at the constructor?

Reply:
Ah good point. Let's do that